librz/core/rop.c

```diff
@@ -0,0 +1,774 @@
+	case RZ_ANALYSIS_OP_TYPE_RET:
+	case RZ_ANALYSIS_OP_TYPE_UCALL:
+	case RZ_ANALYSIS_OP_TYPE_RCALL:
+	case RZ_ANALYSIS_OP_TYPE_ICALL:
+	case RZ_ANALYSIS_OP_TYPE_IRCALL:
+	case RZ_ANALYSIS_OP_TYPE_UJMP:
+	case RZ_ANALYSIS_OP_TYPE_RJMP:
+	case RZ_ANALYSIS_OP_TYPE_IJMP:
+	case RZ_ANALYSIS_OP_TYPE_IRJMP:
+	case RZ_ANALYSIS_OP_TYPE_JMP:
+	case RZ_ANALYSIS_OP_TYPE_CALL:
+		return true;
+	}
+	if (crop) { // if conditional jumps, calls and returns should be used for the gadget-search too
+		switch (aop->type) {
+		case RZ_ANALYSIS_OP_TYPE_CJMP:
+		case RZ_ANALYSIS_OP_TYPE_UCJMP:
+		case RZ_ANALYSIS_OP_TYPE_CCALL:
+		case RZ_ANALYSIS_OP_TYPE_UCCALL:
+		case RZ_ANALYSIS_OP_TYPE_CRET:
+			return true;
+		}
+	}
+	return false;
+}
+
+static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, RzCmdStateOutput *state) {
+	RzCoreAsmHit *hit = NULL;
+	RzListIter *iter;
+	RzList *ropList = NULL;
+	unsigned int size = 0;
+	char *asmop_str = NULL, *asmop_hex_str = NULL;
+	RzAnalysisOp aop = RZ_EMPTY;
+	const char *comment = NULL;
+	Sdb *db = NULL;
+	const bool colorize = rz_config_get_i(core->config, "scr.color");
+	const bool rop_comments = rz_config_get_i(core->config, "rop.comments");
+	const bool esil = rz_config_get_i(core->config, "asm.esil");
+	const bool rop_db = rz_config_get_i(core->config, "rop.db");
+	ut8 *buf = NULL;
+	RzStrBuf *colored_asm = NULL, *bw_str = NULL;
+	if (rop_db) {
+		db = sdb_ns(core->sdb, "rop", true);
+		ropList = rz_list_newf(free);
+		if (!db) {
+			RZ_LOG_ERROR("core: Could not create SDB 'rop' namespace\n");
+			rz_list_free(ropList);
+			return;
+		}
+	}
+
+	rz_cmd_state_output_set_columnsf(state, "XXs", "addr", "bytes", "disasm");
+	if (state->mode == RZ_OUTPUT_MODE_JSON) {
+		pj_o(state->d.pj);
+		pj_ka(state->d.pj, "opcodes");
+	} else if (state->mode == RZ_OUTPUT_MODE_QUIET) {
+		rz_cons_printf("0x%08" PFMT64x ":", ((RzCoreAsmHit *)rz_list_first(hitlist))->addr);
+	}
+	const ut64 addr = ((RzCoreAsmHit *)rz_list_first(hitlist))->addr;
+
+	rz_list_foreach (hitlist, iter, hit) {
+		RzAsmOp *asmop = rz_asm_op_new();
+		switch (state->mode) {
+		case RZ_OUTPUT_MODE_JSON:
+			buf = malloc(hit->len);
+			if (!buf) {
+				goto cleanup;
+			}
+			rz_io_read_at(core->io, hit->addr, buf, hit->len);
+			rz_asm_set_pc(core->rasm, hit->addr);
+			rz_asm_disassemble(core->rasm, asmop, buf, hit->len);
+			rz_analysis_op_init(&aop);
+			rz_analysis_op(core->analysis, &aop, hit->addr, buf, hit->len, RZ_ANALYSIS_OP_MASK_ESIL);
+			size += hit->len;
+			if (aop.type != RZ_ANALYSIS_OP_TYPE_RET) {
+				char *opstr_n = rz_str_newf(" %s", RZ_STRBUF_SAFEGET(&aop.esil));
+				rz_list_append(ropList, opstr_n);
+			}
+			pj_o(state->d.pj);
+			pj_kn(state->d.pj, "offset", hit->addr);
+			pj_ki(state->d.pj, "size", hit->len);
+			pj_ks(state->d.pj, "opcode", rz_asm_op_get_asm(asmop));
+			pj_ks(state->d.pj, "type", rz_analysis_optype_to_string(aop.type));
+			pj_end(state->d.pj);
+			free(buf);
+			rz_analysis_op_fini(&aop);
```

Comment:
this is a lot of code within the switch. probably is better to make a static function which handles only this form of printing.